src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2Serializer.cs

```diff
@@ -82,6 +88,200 @@ internal static bool CanCreateValidUri(string uriString, UriKind uriKind)
             return Uri.TryCreate(uriString, uriKind, out Uri tempUri);
         }
 
+        private Saml2EncryptedAssertion EncryptAssertion(byte[] assertionData, EncryptingCredentials encryptingCredentials)
+        {
+            ValidateEncryptingCredentials(encryptingCredentials);
+
+            var encryptedAssertion = new Saml2EncryptedAssertion();
+
+            // SymmetricSecurityKey is provided:
+            // Pre-shared symmetric key (session key) is used to encrypt an assertion
+            // Session key will not be serialized, but KeyName will be set if available
+            if (encryptingCredentials.Key is SymmetricSecurityKey)
+            {
+                encryptedAssertion.EncryptedData = CreateEncryptedDataHelper((SymmetricSecurityKey)encryptingCredentials.Key, encryptingCredentials.Enc, assertionData, encryptingCredentials);
+                encryptedAssertion.EncryptedData.KeyInfo.KeyName = encryptingCredentials.Key.KeyId;
+            }
+            // AsymmetricSecurityKey is provided:
+            // New session key will be created to encrypt an assertion
+            // Session key will be wrapped with provided AsymmetricSecurityKey
+            else if (encryptingCredentials.Key is AsymmetricSecurityKey)
+            {
+                SymmetricSecurityKey sessionKey = CreateSessionKeyBytes(encryptingCredentials.Enc);
+
+                // Create EncryptedData
+                encryptedAssertion.EncryptedData = CreateEncryptedDataHelper(sessionKey, encryptingCredentials.Enc, assertionData, encryptingCredentials);
+                // Set EncryptedData RetrievalMethodUri to ID of EncryptedKey
+                encryptedAssertion.EncryptedData.KeyInfo.RetrievalMethodUri = encryptingCredentials.Key.KeyId;
+
+                // Create EncryptedKey
+                encryptedAssertion.EncryptedKey = CreateEncryptedKeyHelper((AsymmetricSecurityKey)encryptingCredentials.Key, sessionKey, encryptingCredentials.Alg, encryptingCredentials);
+                // Add reference from EncrypedKey to EncryptedData
+                encryptedAssertion.EncryptedKey.AddReference(new DataReference(encryptedAssertion.EncryptedData.Id));
+
+                // Set Digest method for EncryptedKey
+                if (SecurityAlgorithms.RsaOaepMgf1pKeyWrap.Equals(encryptingCredentials.Alg, StringComparison.Ordinal))
+                    encryptedAssertion.EncryptedKey.EncryptionMethod.DigestMethod = SecurityAlgorithms.Sha1Digest;
+
+                // Set X509CertificateData
+                if (encryptingCredentials.Key is X509SecurityKey)
```

Comment:
> if (encryptingCredentials.Key is X509SecurityKey) [](start = 15, length = 50)

same though as above, what if this is not true, will the method still behave as expected? what about dependencies upstream? consider if an exception should be thrown.

Reply:
this element is optional, but it's might be useful when when users are providing custom KeyResolving logic using TokenDecryptionKeyResolver delegate.
Current implementation is relying on using TokenDecryptionKey (set on TokenValidationParams) to decrypt an assertion, if TokenDecryptionKeyResolver delegate is not set.